src/okareo/crewai_logger.py

```diff
@@ -0,0 +1,275 @@
+
+                raise
+
+    return wrapper
+
+
+class LiteLLMInstrumentation(BaseInstrumentor):  # type: ignore
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return ["litellm >= 1.48.0", "opentelemetry-api >= 1.0.0"]
+
+    def _instrument(self, **kwargs: Any) -> None:
+        tracer_provider: Optional[TracerProvider] = kwargs.get("tracer_provider")
+        tracer = get_tracer(__name__, "", tracer_provider)
+        version: str = importlib.metadata.version("litellm")
+        wrap_function_wrapper(
+            "litellm",
+            "completion",
+            completion_wrapper(version, tracer),
+        )
+
+        wrap_function_wrapper(
+            "litellm",
+            "text_completion",
+            completion_wrapper(version, tracer),
+        )
+
+    def _uninstrument(self, **kwargs: Any) -> None:
+        pass
+
+def chat_completions_create(version: Any, tracer: Any) -> Any:
+    def wrapper(wrapped: Any, instance: Any, args: Any, kwargs: Any) -> Any:
+        with tracer.start_as_current_span("openai.chat.completions.create") as span:
+            span.set_attribute("openai.version", version)
+
+            model = kwargs.get("model")
+            if model:
+                span.set_attribute("openai.model", model)
+
+
+            messages = kwargs.get("messages")
+            span.set_attribute("openai.message_get", json.dumps(messages))
+
+            try:
+                result = wrapped(*args, **kwargs)
+
+                if hasattr(result, "usage"):
+                    span.set_attribute(
+                        "openai.completion_tokens", result.usage.completion_tokens
+                    )
+                    span.set_attribute(
+                        "openai.prompt_tokens", result.usage.prompt_tokens
+                    )
+                    span.set_attribute("openai.total_tokens", result.usage.total_tokens)
+
+                return result
+            except Exception as e:
+                span.record_exception(e)
+                raise
+
+    return wrapper
+
+
+class OpenAIInstrumentation(BaseInstrumentor):  # type: ignore
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return ["openai >= 0.27.0", "opentelemetry-api >= 1.0.0"]
+
+    def _instrument(self, **kwargs: Any) -> None:
+        tracer_provider: Optional[TracerProvider] = kwargs.get("tracer_provider")
+        tracer = get_tracer(__name__, "", tracer_provider)
+        version: str = importlib.metadata.version("openai")
+
+        wrap_function_wrapper(
+            "openai.resources.chat.completions",
+            "Completions.create",
+            chat_completions_create(version, tracer),
+        )
+        wrap_function_wrapper(
+            "openai.resources.chat.completions",
+            "AsyncCompletions.create",
+            chat_completions_create(version, tracer),
+        )
+
+    def _uninstrument(self, **kwargs: Any) -> None:
+        pass
+
+
+class CrewAILogger:
```

Comment:
if we have the option of python contextmanager for Autogen, should we not also do the same for Crew?

Also for autogen i've commented to have explicity start/end as an option in addition to contextmanager, making a note so we could keep it consistent